iox_query/src/exec/cross_rt_stream.rs

```diff
@@ -0,0 +1,358 @@
+    where
+        S: Stream<Item = Result<X, E>> + Send + 'static,
+        C: Fn(executor::Error) -> E + Send + 'static,
+    {
+        Self::new_with_tx(|tx| {
+            // future to be run in the other runtime
+            let tx_captured = tx.clone();
+            let fut = async move {
+                tokio::pin!(stream);
+
+                while let Some(res) = stream.next().await {
+                    if tx_captured.send(res).await.is_err() {
+                        // receiver gone
+                        return;
+                    }
+                }
+            };
+
+            // future for this runtime (likely the tokio/tonic/web driver)
+            async move {
+                if let Err(e) = exec.spawn(fut).await {
+                    let e = converter(e);
+
+                    // last message, so we don't care about the receiver side
+                    tx.send(Err(e)).await.ok();
+                }
+            }
+        })
+    }
+}
+
+impl<X> CrossRtStream<Result<X, DataFusionError>>
+where
+    X: Send + 'static,
+{
+    /// Create new stream based on an existing stream that transports [`Result`]s w/ [`DataFusionError`]s.
+    ///
+    /// Also receives an executor that actually executes the underlying stream.
+    pub fn new_with_df_error_stream<S>(stream: S, exec: DedicatedExecutor) -> Self
+    where
+        S: Stream<Item = Result<X, DataFusionError>> + Send + 'static,
+    {
+        Self::new_with_error_stream(stream, exec, |e| {
+            DataFusionError::Context(
+                "Join Error".to_string(),
+                Box::new(DataFusionError::External(e.into())),
+            )
+        })
+    }
+}
+
+impl<X> CrossRtStream<Result<X, ArrowError>>
+where
+    X: Send + 'static,
+{
+    /// Create new stream based on an existing stream that transports [`Result`]s w/ [`ArrowError`]s.
+    ///
+    /// Also receives an executor that actually executes the underlying stream.
+    pub fn new_with_arrow_error_stream<S>(stream: S, exec: DedicatedExecutor) -> Self
+    where
+        S: Stream<Item = Result<X, ArrowError>> + Send + 'static,
+    {
+        Self::new_with_error_stream(stream, exec, |e| ArrowError::ExternalError(e.into()))
+    }
+}
+
+impl<T> Stream for CrossRtStream<T> {
+    type Item = T;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = &mut *self;
+
+        if !this.driver_ready {
+            let res = this.driver.poll_unpin(cx);
+
+            if res.is_ready() {
+                this.driver_ready = true;
+            }
+        }
+
+        if this.inner_done {
+            if this.driver_ready {
+                Poll::Ready(None)
+            } else {
+                Poll::Pending
+            }
+        } else {
+            let res = this.inner.poll_next_unpin(cx);
+
+            match res {
+                Poll::Pending => Poll::Pending,
```

Comment:
I think you can use `ready!` here:

```suggestion
            match ready!(his.inner.poll_next_unpin(cx)) {
                Some(x)=> Poll::Ready(Some(x)),
                None => {..}
             }
```